classes/choice_importer.php

```diff
@@ -0,0 +1,305 @@
+        $importstatus->errors = array();
+        $importstatus->importcount = 0;
+        $importstatus->rowcount = 1;  // Start at 1 for header.
+
+        $importstatus->readcount = $reader->load_csv_content($content, 'UTF-8', 'comma');
+        $importerror = $reader->get_error();
+
+        if ($importerror) {
+            $importstatus->status = self::IMPORT_STATUS_SETUP_ERROR;
+            $importstatus->errors[] = $importerror;
+        } else {
+            if (empty($importstatus->readcount)) {
+                $importstatus->status = self::IMPORT_STATUS_DATA_ERROR;
+                $importstatus->errors[] = get_string('csvempty', 'ratingallocate');
+            } else {
+                // Trim whitespace on headers, just in case.
+                if (!$fieldnames = array_map('trim', $reader->get_columns())) {
+                    throw new moodle_exception('cannotreadtmpfile', 'error');
+                }
+
+                // Check for any missing required fields.
+                $missingfields = array_diff(self::REQUIRED_FIELDS, $fieldnames);
+                if ($missingfields) {
+                    $importstatus->status = self::IMPORT_STATUS_SETUP_ERROR;
+                    $importstatus->errors[] = get_string('csvupload_missing_fields', 'ratingallocate',
+                        self::print_fields($missingfields));
+                    return $importstatus;
+                }
+
+                // Map group names to group IDs.
+                $allgroups = $this->ratingallocate->get_group_selections();
+                $groupidmap = array_flip($allgroups);
+
+                // Start DB transaction.
+                $transaction = $this->ratingallocate->db->start_delegated_transaction();
+                try {
+
+                    $reader->init();
+                    while ($record = $reader->next()) {
+                        $importstatus->rowcount++;
+                        $recordmap = new \stdClass();
+
+                        // Map cell contents to field names.
+                        foreach ($record as $col => $cell) {
+                            $fieldname = $fieldnames[$col];
+
+                            // Skip non-required field columns.
+                            if (!in_array($fieldname, self::REQUIRED_FIELDS)) {
+                                continue;
+                            }
+
+                            if ($fieldname == 'groups') {
+                                $groups = array();
+
+                                // Turn off 'usegroups' if no groups specified.
+                                if (empty($cell)) {
+                                    $recordmap->usegroups = false;
+                                } else {
+                                    // For fault tolerance, trim any surrounding whitespace.
+                                    $cell = trim($cell);
+                                    $parts = explode(',', $cell);
```

Comment:
Need to use a different delimiter here, e. g. ";", as comma is already used as csv delimiter

Reply:
thanks for the review - as mentioned in #224 I'd prefer to wait for the plugin maintainers to provide feedback on any changes they would like to see before we come back to this again.

Reply:
I agree, otherwise it is not possible to specify more than one group as the comma is interpreted as another column for the csv import.

Reply:
I didn't know about the quotation marks, thanks for pointing that out! This underlines my point that the desired formatting for the CSV File should be communicated clearly  to the user e.g by examples. However, it may still be more intuitive to use a different separator here so that the quotation marks are not necessary for multiple groups.